### PR TITLE
Support Search Template API Explain Query String Argument

### LINKED ISSUE #83363: Search Template API Explain Query String Argument Ignored
### Elasticsearch Version

7.16.3

### Installed Plugins

_No response_

### Java Version

_bundled_

### OS Version

Windows 10

### Problem Description

I'm looking into a report of a [bug raised against the .NET client](https://github.com/elastic/elasticsearch-net/issues/5040) for use of search templates. The root cause seems to be that while the [REST API spec](https://github.com/elastic/elasticsearch-net/blob/8.0/src/ApiGenerator/RestSpecification/Core/search_template.json#L82) and [docs](https://github.com/elastic/elasticsearch-net/issues/5040) mention that "explain" can be included on the query string, doing so doesn't result in an explanation being returned. Only if it's included in the body is the explanation returned. This appears to be a bug with the parsing of search template rest requests as the `explain` parameter is not retrieved.

### Steps to Reproduce

Make a request with "explain=true" on the query string:

```
GET test/_search/template?explain=true
{
 "id": "my-search-template",
 "params": {
    "query_string": "hello world",
    "from": 0,
    "size": 10
  }
}
```

**Expected result:**
The explanation is included in the results.

**Actual:** 
The explanation is not included.

### Logs (if relevant)

_No response_

### PR DESCRIPTION
**Pull request by jackpan123**
Closes #83363

### CODE REVIEW

**Comment by benwtrent** [comment]
@jackpan123 I like this change!

Could you add a test? Probably to `SearchTemplateIT` that ensures `explain` executes? This way we don't accidentally break it in the future :)

**Comment by jackpan123** [comment]
@benwtrent I have a problem where I am trying to add the 'testTemplateWithExplainQueryStringArgument' method into the 'SearchTemplateIT' class, but I am unsure of how to set the query string argument. If I call 'SearchTemplateRequestBuilder#setExplain' directly, it will not test my case. Could you give me some guidance? My test code looks like the following:
```
public void testTemplateWithExplainQueryStringArgument() throws Exception {
        SearchRequest searchRequest = new SearchRequest();
        searchRequest.indices("test");
        String query = """
            { "query": {"match_all": {}}, "size" : "{{my_size}}" }""";
        SearchTemplateResponse searchResponse = new SearchTemplateRequestBuilder(client()).setRequest(searchRequest)
            .setScript(query)
            .setScriptType(ScriptType.INLINE)
            .setScriptParams(Collections.singletonMap("my_size", 1))
            .setExplain(true)
            .get();

        assertThat(searchResponse.toString().contains("_explanation"), equalTo(true));

    }
```

**Comment by jackpan123** [comment]
@benwtrent I tried another test class called `RestSearchTemplateActionTests`. It allows me to add `explain` as a query string argument, but I am unsure how to retrieve the response for assertion. My test code is as follows:
```
public void testExplainInPath() {
      Map<String, String> params = new HashMap<>();
      params.put("explain", "true");

      RestRequest request = new FakeRestRequest.Builder(xContentRegistry()).withHeaders(
          Map.of("Content-Type", contentTypeHeader, "Accept", contentTypeHeader)
      ).withMethod(RestRequest.Method.GET).withPath("/some_index/_search/template").withParams(params).build();

      dispatchRequest(request);
  }
```

File: modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/RestSearchTemplateAction.java [hunk 1]
```diff
@@ -72,12 +72,22 @@ public RestChannelConsumer prepareRequest(RestRequest request, NodeClient client
             searchTemplateRequest = SearchTemplateRequest.fromXContent(parser);
         }
         searchTemplateRequest.setRequest(searchRequest);
-
+        prepareRequest(searchTemplateRequest, request);
```
Comment (benwtrent) [inline modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/RestSearchTemplateAction.java:75]:
```suggestion
        // This param is parsed within the search request
        if (searchRequest.source().explain() != null) {
            searchTemplateRequest.setExplain(searchRequest.source().explain());
        }
```
We should take the parameter from there as it is already parsed, and check if its supplied.

**Review by benwtrent — CHANGES_REQUESTED** [review]
Please add a yamlRestTest to `modules/lang-mustache/src/yamlRestTest/resources/rest-api-spec/test/lang_mustache/30_search_template.yml`

Here is what I think the test should look like. 

```yaml
---
"Test explain templated search":
  - skip:
      version: " - 8.7.99"
      reason: Bug fix for explain in search template added in 8.8.0

  - do:
      index:
        index:  test
        id:     "1"
        body:   { "text": "value1" }
  - do:
      index:
        index:  test
        id:     "2"
        body:   { "text": "value2" }
  - do:
      indices.refresh: {}

  - do:
      search_template:
        explain: true
        body: { "source" : { "query": { "term": { "text": { "value": "{{template}}" } } } }, "params": { "template": "value1" } }

  - is_true: hits.hits.0._explanation
 # Test consistency with _search API, where query parameter takes priority over body
  - do:
      search_template:
        explain: true
        body: { "explain": false, "source" : { "query": { "term": { "text": { "value": "{{template}}" } } } }, "params": { "template": "value1" } }

  - is_true: hits.hits.0._explanation

  - do:
      search_template:
        explain: false
        body: {"explain": true, "source" : { "query": { "term": { "text": { "value": "{{template}}" } } } }, "params": { "template": "value1" } }

  - is_false: hits.hits.0._explanation

```

**Comment by jackpan123** [comment]
@benwtrent Thanks! I already add the test!

File: modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/RestSearchTemplateAction.java [hunk 2]
```diff
@@ -72,12 +72,25 @@ public RestChannelConsumer prepareRequest(RestRequest request, NodeClient client
...
+
+    /**
+     * Parses the rest request on top of the SearchTemplateRequest, preserving
+     * values that are not overridden by the rest request.
+     */
+    private static void prepareRequest(final SearchTemplateRequest searchTemplateRequest, RestRequest request) {
+        if (request.hasParam("explain")) {
+            searchTemplateRequest.setExplain(request.paramAsBoolean("explain", false));
+        }
+    }
```
Comment (benwtrent) [inline modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/RestSearchTemplateAction.java:95]:
```suggestion
```

No longer used.

**Review by benwtrent — COMMENTED** [review]
one more thing. Its looking good.

**Comment by benwtrent** [comment]
jenkins retest this please

**Comment by jackpan123** [comment]
@benwtrent I have a question: if I need to add more arguments to the query string, it would be more flexible to use a method for handling them, such as the `RestSearchAction#parseSearchSource` method.

**Comment by benwtrent** [comment]
@jackpan123 

> if I need to add more arguments to the query string

Do you mean new URL parameters? It honestly depends on the parameter. I don't know of any new parameters we need or should add to template search.

`explain` is a unique case as it was already "supported" just not unified in usage with search.

**Comment by benwtrent** [comment]
@elasticmachine update branch

**Comment by benwtrent** [comment]
jenkins retest this please

**Review by benwtrent — APPROVED** [review]

**Comment by jackpan123** [comment]
> @jackpan123
> 
> > if I need to add more arguments to the query string
> 
> Do you mean new URL parameters? It honestly depends on the parameter. I don't know of any new parameters we need or should add to template search.
> 
> `explain` is a unique case as it was already "supported" just not unified in usage with search.

I got it now. Thanks for your patient explanation.

**Pull request merged**